File: THCS/2022_2023/NgheAn_Vinh_2223/Bai3/MINPLACES.py
def min_parking_spaces(arrival, departure):
    # Sắp xếp thời gian đến và thời gian rời bến
    arrival.sort()
    departure.sort()

    # Biến đếm số chỗ đỗ xe hiện tại và tối đa
    spaces_needed = 0
    max_spaces = 0

    # Chỉ số cho thời gian đến và đi
    i, j = 0, 0
    n = len(arrival)

    # Duyệt qua tất cả các xe
    while i < n and j < n:
        if arrival[i] < departure[j]:
            # Nếu có xe đến trước khi xe rời đi, tăng số chỗ đỗ cần thiết
            spaces_needed += 1
            max_spaces = max(max_spaces, spaces_needed)
            i += 1
        else:
            # Nếu có xe rời đi, giảm số chỗ đỗ cần thiết
            spaces_needed -= 1
            j += 1

    return max_spaces

def main():
    with open("MINPLACES.INP", "r") as input_file:
        # Đọc số bộ dữ liệu
        T = int(input_file.readline().strip())
        results = []

        for _ in range(T):
            # Đọc số xe trong bộ dữ liệu
            N = int(input_file.readline().strip())

            # Đọc danh sách thời gian đến và rời bến
            arrival = list(map(int, input_file.readline().strip().split()))
            departure = list(map(int, input_file.readline().strip().split()))

            # Tính số chỗ đỗ xe tối thiểu cho bộ dữ liệu này
            result = min_parking_spaces(arrival, departure)
            results.append(result)

    # Ghi kết quả ra file MINPLACES.OUT
    with open("MINPLACES.OUT", "w") as output_file:
        for result in results:
            output_file.write(f"{result}\n")

if __name__ == "__main__":
    main()

File: THCS/2022_2023/NgheAn_Vinh_2223/Bai3/MINPLACES.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
using namespace std;

// Hàm tính số chỗ đỗ xe tối thiểu
int min_parking_spaces(vector<int>& arrival, vector<int>& departure, int n) {
    // Sắp xếp thời gian đến và rời bến
    sort(arrival.begin(), arrival.end());
    sort(departure.begin(), departure.end());

    int spaces_needed = 0;  // Số chỗ đỗ xe hiện tại
    int max_spaces = 0;     // Số chỗ đỗ tối đa
    int i = 0, j = 0;       // Chỉ số cho thời gian đến và đi

    // Duyệt qua các xe
    while (i < n && j < n) {
        if (arrival[i] < departure[j]) {
            // Nếu có xe đến trước khi xe rời đi, tăng số chỗ đỗ cần thiết
            spaces_needed++;
            max_spaces = max(max_spaces, spaces_needed);
            i++;
        } else {
            // Nếu có xe rời đi, giảm số chỗ đỗ cần thiết
            spaces_needed--;
            j++;
        }
    }

    return max_spaces;
}

int main() {
    ifstream infile("MINPLACES.INP");
    ofstream outfile("MINPLACES.OUT");

    int T;
    infile >> T;  // Đọc số bộ dữ liệu

    while (T--) {
        int N;
        infile >> N;  // Đọc số xe trong bộ dữ liệu

        vector<int> arrival(N), departure(N);

        // Đọc danh sách thời gian đến
        for (int i = 0; i < N; ++i) {
            infile >> arrival[i];
        }

        // Đọc danh sách thời gian rời bến
        for (int i = 0; i < N; ++i) {
            infile >> departure[i];
        }

        // Tính số chỗ đỗ xe tối thiểu
        int result = min_parking_spaces(arrival, departure, N);
        outfile << result << endl;
    }

    infile.close();
    outfile.close();

    return 0;
}
